File: src/LAB2/i.py
class Node:
    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add_front(self, value):
        node = Node(value)
        if not self.head:
            self.head = self.tail = node
        else:
            node.next = self.head
            self.head.prev = node
            self.head = node
        print("ok")

    def add_back(self, value):
        node = Node(value)
        if not self.tail:
            self.head = self.tail = node
        else:
            node.prev = self.tail
            self.tail.next = node
            self.tail = node
        print("ok")

    def erase_front(self):
        if not self.head:
            print("error")
            return
        val = self.head.value
        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = None
        print(val)

    def erase_back(self):
        if not self.tail:
            print("error")
            return
        val = self.tail.value
        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None
        print(val)

    def front(self):
        if not self.head:
            print("error")
        else:
            print(self.head.value)

    def back(self):
        if not self.tail:
            print("error")
        else:
            print(self.tail.value)

    def clear(self):
        self.head = self.tail = None
        print("ok")

dll = DoublyLinkedList()
while True:
    try:
        cmd = input().strip().split()
    except EOFError:
        break
    if not cmd:
        continue
    if cmd[0] == "add_front":
        dll.add_front(cmd[1])
    elif cmd[0] == "add_back":
        dll.add_back(cmd[1])
    elif cmd[0] == "erase_front":
        dll.erase_front()
    elif cmd[0] == "erase_back":
        dll.erase_back()
    elif cmd[0] == "front":
        dll.front()
    elif cmd[0] == "back":
        dll.back()
    elif cmd[0] == "clear":
        dll.clear()
    elif cmd[0] == "exit":
        print("goodbye")
        break

File: src/LAB2/i.cpp
#include <iostream>
#include <string>
using namespace std;

struct Node {
    string data;
    Node* prev;
    Node* next;
    Node(string d) {
        data = d;
        prev = NULL;
        next = NULL;
    }
};

int main() {
    Node* head = NULL;
    Node* tail = NULL;
    string command;

    while (cin >> command) {
        if (command == "add_front") {
            string book;
            cin >> book;
            Node* node = new Node(book);
            if (!head) {
                head = tail = node;
            } else {
                node->next = head;
                head->prev = node;
                head = node;
            }
            cout << "ok" << endl;
        } else if (command == "add_back") {
            string book;
            cin >> book;
            Node* node = new Node(book);
            if (!tail) {
                head = tail = node;
            } else {
                tail->next = node;
                node->prev = tail;
                tail = node;
            }
            cout << "ok" << endl;
        } else if (command == "erase_front") {
            if (!head) {
                cout << "error" << endl;
            } else {
                Node* temp = head;
                cout << temp->data << endl;
                head = head->next;
                if (head) head->prev = NULL;
                else tail = NULL;
                delete temp;
            }
        } else if (command == "erase_back") {
            if (!tail) {
                cout << "error" << endl;
            } else {
                Node* temp = tail;
                cout << temp->data << endl;
                tail = tail->prev;
                if (tail) tail->next = NULL;
                else head = NULL;
                delete temp;
            }
        } else if (command == "front") {
            if (!head) cout << "error" << endl;
            else cout << head->data << endl;
        } else if (command == "back") {
            if (!tail) cout << "error" << endl;
            else cout << tail->data << endl;
        } else if (command == "clear") {
            while (head) {
                Node* temp = head;
                head = head->next;
                delete temp;
            }
            tail = NULL;
            cout << "ok" << endl;
        } else if (command == "exit") {
            cout << "goodbye" << endl;
            break;
        }
    }
    return 0;
}